src/installed.rs

```diff
@@ -0,0 +1,357 @@
+                        let listening = server.handle(InstalledFlowHandler {
+                            auth_code_snd: Mutex::new(tx),
+                        });
+
+                        match listening {
+                            Result::Err(_) => default,
+                            Result::Ok(listening) => {
+                                InstalledFlow {
+                                    client: default.client,
+                                    server: Some(listening),
+                                    port: Some(port),
+                                    auth_code_rcv: Some(rx),
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Handles the token request flow; it consists of the following steps:
+    /// . Obtain a auhorization code with user cooperation or internal redirect.
+    /// . Obtain a token and refresh token using that code.
+    /// . Return that token
+    ///
+    /// It's recommended not to use the DefaultAuthenticatorDelegate, but a specialized one.
+    pub fn obtain_token<'a, AD: AuthenticatorDelegate, S, T>(&mut self,
+                                                             auth_delegate: &mut AD,
+                                                             appsecret: &ApplicationSecret,
+                                                             scopes: S)
+                                                             -> Result<Token, Box<Error>>
+        where T: AsRef<str> + 'a,
+              S: Iterator<Item = &'a T>
+    {
+        use std::error::Error;
+        let authcode = try!(self.get_authorization_code(auth_delegate, &appsecret, scopes));
+        let tokens = try!(self.request_token(&appsecret, &authcode));
+
+        // Successful response
+        if tokens.access_token.is_some() {
+            let token = Token {
+                access_token: tokens.access_token.unwrap(),
+                refresh_token: tokens.refresh_token.unwrap(),
+                token_type: tokens.token_type.unwrap(),
+                expires_in: tokens.expires_in,
+                expires_in_timestamp: None,
+            };
+
+            Result::Ok(token)
+        } else {
+            let err = io::Error::new(io::ErrorKind::Other,
+                                     format!("Token API error: {} {}",
+                                             tokens.error.unwrap_or("<unknown err>".to_string()),
+                                             tokens.error_description
+                                                   .unwrap_or("".to_string()))
+                                         .as_str());
+            Result::Err(Box::new(err))
+        }
+    }
+
+    /// Obtains an authorization code either interactively or via HTTP redirect (see
+    /// InstalledFlowReturnMethod).
+    /// TODO(dermesser): Add timeout configurability!
```

Comment:
I don't have a good solution in mind right now for implementing a timeout for stdin. I'll remove the line and create an issue.